applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py

```diff
@@ -0,0 +1,238 @@
+            "domain_size": -1,
+            "model_import_settings": {
+                "input_type": "mdpa",
+                "input_filename": "unknown_name",
+                "reorder": false
+            },
+            "distance_reading_settings"    : {
+                "import_mode"         : "from_mdpa",
+                "distance_file_name"  : "no_distance_file"
+            },
+            "maximum_iterations": 7,
+            "echo_level": 0,
+            "consider_periodic_conditions"	: false,
+            "time_order": 2,
+            "compute_reactions": false,
+            "reform_dofs_at_each_step": false,
+            "relative_velocity_tolerance": 1e-3,
+            "absolute_velocity_tolerance": 1e-5,
+            "relative_pressure_tolerance": 1e-3,
+            "absolute_pressure_tolerance": 1e-5,
+            "linear_solver_settings":   {
+                "solver_type": "AmgclMPISolver",
+                "tolerance": 1.0e-6,
+                "max_iteration": 10,
+                "scaling": false,
+                "verbosity": 0,
+                "preconditioner_type": "None",
+                "krylov_type": "cg"
+            },
+            "volume_model_part_name" : "volume_model_part",
+            "skin_parts": [""],
+            "no_skin_parts":[""],
+            "time_stepping": {
+                "automatic_time_step" : true,
+                "CFL_number"          : 1,
+                "minimum_delta_time"  : 1e-4,
+                "maximum_delta_time"  : 0.01
+            },
+            "move_mesh_flag": false,
+            "formulation": {
+                "dynamic_tau": 1.0
+            }
+        }""")
+
+        settings.ValidateAndAssignDefaults(default_settings)
+        return settings
+
+    def __init__(self, model, custom_settings):
+        # (!!!) Note: deliberately calling the constructor of the base python solver (the parent of my parent)
+        super(NavierStokesMPITwoFluidsSolver, self).__init__(model,custom_settings)
+
+        self.element_name = "TwoFluidNavierStokes"
+        self.condition_name = "NavierStokesWallCondition"
+        self.min_buffer_size = 3
+
+        self._is_printing_rank = (KratosMPI.mpi.rank == 0)
+
+        if (self.settings["solver_type"].GetString() == "TwoFluids"):
+            self.element_name = "TwoFluidNavierStokes"
+
+        ## Construct the linear solver
+        import trilinos_linear_solver_factory
+        self.trilinos_linear_solver = trilinos_linear_solver_factory.ConstructSolver(self.settings["linear_solver_settings"])
+
+        if self._IsPrintingRank():
+            KratosMultiphysics.Logger.PrintInfo("NavierStokesMPITwoFluidsSolver","Construction of NavierStokesMPITwoFluidsSolver finished.")
+
+
+    def AddVariables(self):
+
+        super(NavierStokesMPITwoFluidsSolver, self).AddVariables()
+        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PARTITION_INDEX)
+
+        KratosMPI.mpi.world.barrier()
```

Comment:
I dont think those barriers are needed
Afaik the other solvers dont have it
Simon did you have problems?

Reply:
For the four processes on my laptop it seems to work without the barrier.
However: A comparable structure concerning the arrangement of barriers is used in TrilinosNavierStokesSolverMonolithic, NavierStokesMPIEmbeddedMonolithicSolver, TrilinosNavierStokesSolverFractionalStep, and AdjointVMSMonolithicMPISolver. Also from a look at the code, I think that leaving them out could cause a lot of trouble. Thus, I believe that one should not rely on a experimental result on a 4-process machine and that the barriers should remain. Though the "but-it-does-not-cost-so-much" argument is not the best, one could consider that this function is not repeated in an innermost loop but only called once for a simulation run.

Reply:
Hm ok

Fine for me, maybe @jcotela knows more here
I‘d be interested to hear his opinion

Reply:
It is true that most MPI fluid solvers have it (probably my fault) but in fact I am not so sure it is needed.

Reply:
Would you agree if we keep the barrier for now, @jcotela @philbucher ? At the same time, I promise that I will keep the question in mind. After merging, the implementation could by tested on the cluster which could give as more reliable results.